benchmark/single-source/Breadcrumbs.swift

```diff
@@ -0,0 +1,499 @@
+    for _ in 0 ..< iterations {
+      for index in inputIndices {
+        blackHole(inputString._toUTF16Offset(index))
+      }
+    }
+  }
+}
+
+/// Split a string into `count` random slices and convert their UTF-16 offsets
+/// into String index ranges.
+class UTF16ToIdxRange: BenchmarkBase {
+  let count: Int
+  var inputOffsets: [Range<Int>] = []
+
+  init(workload: Workload, count: Int) {
+    self.count = count
+    super.init(name: "Breadcrumbs.UTF16ToIdxRange", workload: workload)
+  }
+
+  override func setUp() {
+    super.setUp()
+    var rng = LCRNG(seed: seed)
+    inputOffsets = (
+      0 ..< inputString.utf16.count
+    ).randomIndexRanges(count: count, using: &rng)
+  }
+
+  override func tearDown() {
+    super.tearDown()
+    inputOffsets = []
+  }
+
+  @inline(never)
+  override func run(iterations: Int) {
+    for _ in 0 ..< iterations {
+      for range in inputOffsets {
+        blackHole(inputString._toUTF16Indices(range))
+      }
+    }
+  }
+}
+
+/// Split a string into `count` random slices and convert their index ranges
+/// into into UTF-16 offset pairs.
+class IdxToUTF16Range: BenchmarkBase {
+  let count: Int
+  var inputIndices: [Range<String.Index>] = []
+
+  init(workload: Workload, count: Int) {
+    self.count = count
+    super.init(name: "Breadcrumbs.IdxToUTF16Range", workload: workload)
+  }
+
+  override func setUp() {
+    super.setUp()
+    var rng = LCRNG(seed: seed)
+    inputIndices = self.inputString.randomIndexRanges(count: count, using: &rng)
+  }
+
+  override func tearDown() {
+    super.tearDown()
+    inputIndices = []
+  }
+
+  @inline(never)
+  override func run(iterations: Int) {
+    for _ in 0 ..< iterations {
+      for range in inputIndices {
+        blackHole(inputString._toUTF16Offsets(range))
+      }
+    }
+  }
+}
+
+
+class CopyUTF16CodeUnits: BenchmarkBase {
+  let count: Int
+  var inputIndices: [Range<Int>] = []
+  var outputBuffer: [UInt16] = []
+
+  init(workload: Workload, count: Int) {
+    self.count = count
+    super.init(name: "Breadcrumbs.CopyUTF16CodeUnits", workload: workload)
+  }
+
+  override func setUp() {
+    super.setUp()
+    var rng = LCRNG(seed: seed)
+    inputIndices = (
+      0 ..< inputString.utf16.count
+    ).randomIndexRanges(count: count, using: &rng)
+    outputBuffer = Array(repeating: 0, count: inputString.utf16.count)
+  }
+
+  override func tearDown() {
+    super.tearDown()
+    inputIndices = []
+  }
+
+  @inline(never)
+  override func run(iterations: Int) {
+    outputBuffer.withUnsafeMutableBufferPointer { buffer in
+      for _ in 0 ..< iterations {
+        for range in inputIndices {
+          inputString._copyUTF16CodeUnits(
+            into: UnsafeMutableBufferPointer(rebasing: buffer[range]),
+            range: range)
+        }
+      }
+    }
+  }
+}
+
+/// This is like `UTF16ToIdx` but appends to the string after every index
+/// conversion. In effect, this tests breadcrumb creation performance.
```

Comment:
This comment now doesn't exactly reflect the add/remove mutations going on.

Also, without knowing details of the breadcrumbs implementation, it's not clear to me how this mutation is affecting it… Does it mean that the memoized UTF16 representations are thrown away on each mutation? If so, can you add such explanation to this comment?

Reply:
Damn… I was too slow :-) These are not super important...